### PR TITLE
Compare page 

### PR DESCRIPTION
**Pull request by amit-webkul**

### CODE REVIEW

File: packages/Webkul/Shop/src/Http/Controllers/API/CartController.php [hunk 1]
```diff
@@ -43,9 +57,21 @@ public function index(): JsonResource
...
+                request()['guest_id'] = 1;
+                request()['guest_id'] = null;
+
+                $cart = Cart::addProduct($productId, request()->all());
+            } else {
+                request()['guest_id'] = null;
+                request()['customer_id'] = $customer->id;
+
+                $cart = Cart::addProduct($productId, request()->all());
+            }
```
Comment (suraj-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/API/CartController.php:74]: Need imrpovement here.

File: packages/Webkul/Customer/src/Models/CompareItem.php [hunk 2]
```diff
@@ -31,4 +32,13 @@ public function customer()
     {
         return $this->belongsTo(CustomerProxy::modelClass(), 'customer_id');
     }
+
+    /**
+     * The product that belong to the compare product.
+     *
+     * @return void
```
Comment (suraj-webkul) [inline packages/Webkul/Customer/src/Models/CompareItem.php:39]: Add belongs to return type  here.

**Review by suraj-webkul — CHANGES_REQUESTED** [review]

File: packages/Webkul/Shop/src/Http/Resources/CompareResource.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+namespace Webkul\Shop\Http\Resources;
+
+use Illuminate\Http\Resources\Json\JsonResource;
+
+class CompareResource extends JsonResource
+{
+    /**
+     * Transform the resource into an array.
+     *
+     * @param  \Illuminate\Http\Request
```
Comment (suraj-webkul) [inline packages/Webkul/Shop/src/Http/Resources/CompareResource.php:12]: write @param \Illuminate\Http\Request  $request

File: packages/Webkul/Shop/src/Resources/views/checkout/cart/index.blade.php [hunk 4]
```diff
@@ -144,10 +146,10 @@ class="text-[16px] font-medium"
                                 @lang('shop::app.checkout.cart.coupon.discount')
                             </p>
 
-                            <p class="text-[16px] font-medium cursor-pointer" v-if="! cart.discount_amount">
+                            <p class="text-[16px] font-medium cursor-pointer" v-if="cart.discount_amount == 0">
```
Comment (suraj-webkul) [inline packages/Webkul/Shop/src/Resources/views/checkout/cart/index.blade.php:149]: Break attributes.

File: packages/Webkul/Shop/src/Resources/views/checkout/cart/mini-cart.blade.php [hunk 5]
```diff
@@ -130,7 +130,7 @@ class="text-[30px] font-semibold"
                 getCart() {
                     this.$axios.get('{{ route('shop.checkout.cart.index') }}')
                         .then(response => {
-                            this.cart = response.data.data;
+                            this.cart = response.data.data[0];
```
Comment (suraj-webkul) [inline packages/Webkul/Shop/src/Resources/views/checkout/cart/mini-cart.blade.php:133]: check is it works.

File: packages/Webkul/Shop/src/Resources/views/compare/index.blade.php [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            app.component("v-compare-item", {
+                template: '#v-compare-item-template',
+
+                data() {
+                    return  {
+                        compareItem: [],
+
+                        products: [],
+
+                        isCustomer: '{{ auth()->guard('customer')->user() ? "true" : "false" }}' == 'true',
```
Comment (suraj-webkul) [inline packages/Webkul/Shop/src/Resources/views/compare/index.blade.php:100]: Optimize this one

File: packages/Webkul/Shop/src/Resources/views/compare/index.blade.php [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                    },
+
+                    remove(productId) {
+                        if (this.isCustomer) {
+                            this.$axios.post('{{ route('shop.customers.compare.destroy') }}', {
+                                    '_method': 'DELETE',
+                                    'product_id': productId,
+                                })
+                                .then(response => {
+                                    this.compareItem = response.data.data;
```
Comment (suraj-webkul) [inline packages/Webkul/Shop/src/Resources/views/compare/index.blade.php:150]: if compareItem contains multiple data then rename it to compareItems

**Review by suraj-webkul — CHANGES_REQUESTED** [review]

File: packages/Webkul/Shop/src/Resources/views/compare/index.blade.php (hunk 6 above)
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Resources/views/compare/index.blade.php]: Fixed

File: packages/Webkul/Shop/src/Resources/views/checkout/cart/index.blade.php (hunk 4 above)
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Resources/views/checkout/cart/index.blade.php]: fixed

File: packages/Webkul/Shop/src/Http/Resources/CompareResource.php (hunk 3 above)
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Resources/CompareResource.php:12]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+                'message'  => trans('shop::app.compare.success'),
+            ]);
+        }
+
+        return new JsonResource([
+            'message'  => trans('shop::app.compare.error'),
+        ]);
+    }
+
+    public function moveCart()
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                ]);
+            }
+        } catch (\Exception $exception) {
+            return new JsonResource([
+                'message'   => $exception->getMessage(),
+            ]);
+        }
+    }
+
+    public function moveToWisthlist()
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+            if (! $product) {
+                return response()->json([
+                    'message'  => trans('shop::app.compare.product-removed'),
+                ]);
+
+            } elseif (
+                (! $product->status)
+                || (! $product->visible_individually)
+            ) {
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+            'message' => trans('shop::app.compare.item-add'),
+        ]);
+    }
+
+    /**
+     * Method for compare items to delete products in comparison.
+     *
+     * @return \Illuminate\Http\Response|\Illuminate\View\View
+     */
+    public function destroy(): JsonResource
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+     */
+    public function store()
+    {
+        $customer = auth()->guard('customer')->user();
+
+        if (! $customer) {
+            $product_id = request()->input('compare_product_id');
+
+            if ($product_id) {
+                foreach ($product_id as $id) {
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+        protected CompareItemRepository $compareItemRepository,
+        protected ProductRepository $productRepository,
+        protected WishlistRepository $wishlistRepository
+    ) {
+    }
+
+    /**
+     * Address route index page.
+     *
+     * @return \Illuminate\View\View
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+use Webkul\Customer\Repositories\WishlistRepository;
+use Webkul\Product\Repositories\ProductRepository;
+use Webkul\Shop\Http\Resources\CompareResource;
+
+class CompareController extends APIController
+{
+    /**
+     * Create a new controller instance.
+     *
+     * @param  Webkul\Customer\Repositories\CompareItemRepository  $compareItemsRepository
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CartController.php [hunk 15]
```diff
@@ -28,13 +28,27 @@ public function __construct(
...
     public function index(): JsonResource
     {
+        $customer = auth()->guard('customer')->user();
+
         Cart::collectTotals();
 
-        $cart = Cart::getCart();
+        if ($cart = Cart::getCart()) {
+            if (! $customer) {
+                $cart = $cart->where('is_guest', '=', 1)->get();
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CartController.php]: fixed

File: packages/Webkul/Customer/src/Models/CompareItem.php (hunk 2 above)
Comment (amit-webkul) [reply packages/Webkul/Customer/src/Models/CompareItem.php]: fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CartController.php (hunk 1 above)
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Controllers/API/CartController.php]: fixed

File: packages/Webkul/Shop/src/Http/Resources/ProductResource.php [hunk 16]
```diff
@@ -24,6 +24,10 @@ public function toArray($request)
             'status'      => $this->status,
             'is_new'      => $this->new,
             'is_featured' => $this->featured,
+            'description' => $this->description,
+            'color'       => $this->color,
+            'size'        => $this->size,
+            'description' => $this->description,
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Http/Resources/ProductResource.php]: Fixed

File: packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+
+        if (! $customer) {
+            $compareProductIds = request()->input('compare_product_ids');
+
+            if ($compareProductIds) {
+                foreach ($compareProductIds as $id) {
+                    $product[] = $this->productRepository->findOnewhere([
+                        'id' => $id,
+                    ]);
+                }
```
Comment (devansh-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/API/CompareController.php:50]: Do it in a single query.

File: packages/Webkul/Shop/src/Resources/views/components/layouts/header/desktop/bottom.blade.php [hunk 18]
```diff
@@ -107,13 +107,11 @@ class="icon-camera text-[22px] absolute top-[12px] right-[12px] flex items-cente
...
 
-        @if ($showWishlist)
-            <div>
-                <a href="{{ route('shop.customers.account.wishlist.index') }}">
-                    <span class="icon-heart text-[24px] inline-block cursor-pointer"></span>
-                </a>
-            </div>
-        @endif
+        <div>
+            <a href="{{ route('shop.compare') }}">
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Resources/views/components/layouts/header/desktop/bottom.blade.php:112]: Fixed

File: packages/Webkul/Shop/src/Resources/views/products/view.blade.php [hunk 19]
```diff
@@ -247,12 +253,54 @@ class="icon-share text-[24px]"></span>Share</div>
...
+                        let updatedItems = [productId];
+
+                        let existingItems = this.getStorageValue('compared_product');
+
+                        if (existingItems) {
+                            if (! existingItems.includes(productId)) {
+                                if (existingItems.indexOf(this.productId) == -1) {
+                                    updatedItems = existingItems.concat(updatedItems);
+
+                                    this.setStorageValue('compared_product', updatedItems);
```
Comment (devansh-webkul) [inline packages/Webkul/Shop/src/Resources/views/products/view.blade.php:278]: compare_items

File: packages/Webkul/Shop/src/Resources/views/products/view.blade.php [hunk 20]
```diff
@@ -247,12 +253,54 @@ class="icon-share text-[24px]"></span>Share</div>
...
+                            }
+                        } else {
+                            this.setStorageValue('compared_product', updatedItems);
+                        }
+                    },
+
+                    setStorageValue(key, value) {
+                        window.localStorage.setItem(key, JSON.stringify(value));
+
+                        return true;
```
Comment (devansh-webkul) [inline packages/Webkul/Shop/src/Resources/views/products/view.blade.php:293]: remove the return statement.

File: packages/Webkul/Shop/src/Resources/views/products/view.blade.php [hunk 21]
```diff
@@ -247,12 +253,54 @@ class="icon-share text-[24px]"></span>Share</div>
...
+                                })
+                                .then(response => {
+                                    alert(response.data.data.message);
+                                })
+                                .catch(error => {});
+
+                            return;
+                        }
+
+                        let updatedItems = [productId];
```
Comment (devansh-webkul) [inline packages/Webkul/Shop/src/Resources/views/products/view.blade.php:269]: check this case..

**Review by devansh-webkul — CHANGES_REQUESTED** [review]
One file change left.

File: packages/Webkul/Shop/src/Resources/views/products/view.blade.php [hunk 22]
```diff
@@ -217,6 +217,12 @@ class="icon-share text-[24px]"></span>Share</div>
                 data() {
                     return {
                         qty: 1,
+
+                        reviews: {},
+
+                        page: 1,
```
[the remark this answers is not in the document]
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Resources/views/products/view.blade.php]: fixed

File: packages/Webkul/Shop/src/Resources/views/products/view.blade.php (hunk 20 above)
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Resources/views/products/view.blade.php]: fixed

File: packages/Webkul/Shop/src/Resources/views/products/view.blade.php (hunk 19 above)
Comment (amit-webkul) [reply packages/Webkul/Shop/src/Resources/views/products/view.blade.php]: fixed

**Review by devansh-webkul — APPROVED** [review]

**Review by jitendra-webkul — APPROVED** [review]

**Pull request merged**